main.nf

```diff
@@ -0,0 +1,581 @@
+if(params.rgpl != 'illumina') summary['BAM Read Group Platform']    = params.rgpl
+if(params.rgpu != 'unit1') summary['BAM Read Group Platform Unit']  = params.rgpu
+if(params.rgsm != 20) summary['BAM Read Group Sample']              = params.rgsm
+summary['Max Memory']       = params.max_memory
+summary['Max CPUs']         = params.max_cpus
+summary['Max Time']         = params.max_time
+summary['Number of cores for makeExamples'] = params.j
+summary['DeepVariant trained data model folder'] = params.modelFolder
+summary['DeepVariant trained data model name'] = params.modelName
+summary['Output dir']       = params.outdir
+summary['Working dir']      = workflow.workDir
+summary['Container Engine'] = workflow.containerEngine
+if(workflow.containerEngine) summary['Container'] = workflow.container
+summary['Current home']     = "$HOME"
+summary['Current user']     = "$USER"
+summary['Current path']     = "$PWD"
+summary['Working dir']      = workflow.workDir
+summary['Output dir']       = params.outdir
+summary['Script dir']       = workflow.projectDir
+summary['Config Profile']   = workflow.profile
+if(workflow.profile == 'awsbatch'){
+   summary['AWS Region'] = params.awsregion
+   summary['AWS Queue'] = params.awsqueue
+}
+if(params.email) summary['E-mail Address'] = params.email
+log.info summary.collect { k,v -> "${k.padRight(15)}: $v" }.join("\n")
+log.info "========================================="
+
+
+def create_workflow_summary(summary) {
+
+    def yaml_file = workDir.resolve('workflow_summary_mqc.yaml')
+    yaml_file.text  = """
+    id: 'nf-core-deepvariant-summary'
+    description: " - this information is collected when the pipeline is started."
+    section_name: 'nf-core/deepvariant Workflow Summary'
+    section_href: 'https://github.com/nf-core/deepvariant'
+    plot_type: 'html'
+    data: |
+        <dl class=\"dl-horizontal\">
+${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
+        </dl>
+    """.stripIndent()
+
+   return yaml_file
+}
+
+
+/********************************************************************
+  process preprocessFASTA
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+
+process preprocessFASTA{
+
+  publishDir "$baseDir/sampleDerivatives"
+
+
+  input:
+  file fasta from fasta
+  file fai from fai
+  file fastagz from fastagz
+  file gzfai from gzfai
+  file gzi from gzi
+  output:
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi") into fastaChannel
+  script:
+  """
+  [[ "${params.fai}"=="nofai" ]] &&  samtools faidx $fasta || echo " fai file of user is used, not created"
```

Comment:
Please don't use bash logic, as this is very difficult to debug and is not very traceable.

Instead, do any required logic in groovy, still in the `script` block, but before the `"""`. Then the static command is printed to `.command.sh` and the nextflow trace / report. This is much easier to work with.

Reply:
**So shall I make a process for each of the input files?**

Changing the input files form this (currently):
```bash
//if user inputs fasta set their reference files otherwise use hg19 as default
if( !(false).equals(params.fasta)){
  fasta=file(params.fasta)
  fai = params.fai ? file(params.fai) : false
  fastagz = params.fai ? file(params.fastagz) : false
  gzfai = params.fai ? file(params.gzfai) : false
  gzi = params.fai ? file(params.gzi) : false
} else {
  params.genome = 'hg19'
}

//set genome options
if ( params.genome ){
  fasta = file( params.genome ? params.genomes[ params.genome ].fasta ?: false : false )
  fai = file( params.genome ? params.genomes[ params.genome ].fai ?: false : false )
  fastagz = file( params.genome ? params.genomes[ params.genome ].fastagz ?: false : false )
  gzfai = file( params.genome ? params.genomes[ params.genome ].gzfai ?: false : false )
  gzi = file( params.genome ? params.genomes[ params.genome ].gzi ?: false : false )
  if( !fasta.exists() ) exit 1, "Fasta file not found: ${params.fasta}"
  if( !fai.exists() ) exit 1, "Fai file not found: ${params.fai}"
  if( !fastagz.exists() ) exit 1, "Fastagz file not found: ${params.fastagz}"
  if( !gzfai.exists() ) exit 1, "gzfai file not found: ${params.gzfai}"
  if( !gzi.exists() ) exit 1, "gzi file not found: ${params.gzi}"
} else if( (false).equals(params.fasta)){
  exit 1, "--fasta \"/path/to/your/genome\"  params is required and was not found! or you can use standard genome versions by typing --genome hg19 or --genome h38"
}
```

To something more like this for each of them using channels:
```nextflow
if( params.fai ){ fai = Channel.fromPath(params.fai)}
```
<br /><br />
And then have a process for each of the reference files eg the `fai`, `fastagz` etc
```nextflow
if( (false).equals(params.fai)){
    process preprocessFAI {
        publishDir "$baseDir/sampleDerivatives"

        input:
        file fasta from fasta
        file fai from fai

        output:
        file(fasta)
        file("${fasta}.fai")

        script:
        """
        samtools faidx $fasta
        """
    }
}

if( (false).equals(params.fastqgz)){
    process preprocessFASTGZ {
        publishDir "$baseDir/sampleDerivatives"

        input:
        file fasta from fasta2

        output:
        file("${fasta}.gz")

        script:
        """
        bgzip -c ${fasta} > ${fasta}.gz
        """
    }
}

if( (false).equals(params.gzi)){
    process preprocessGZI {
      ...
    }
}

if( (false).equals(params.gzfai)){
    process preprocessGZFAI {
      ...
    }
}

fastaChannel = Channel.collect()
```

I know there would be a benefit of it being more traceable but this seems far more verbose to me. Is this what you had in mind and do you think it would be better?

Reply:
Yes, this is what I had in mind. It is verbose, and there are more concise ways to write the same thing, but personally I think verbose and readable is much better than concise and confusing.

NB: `if( (false).equals(params.gzfai))` is better written as just `if(!params.gzfai)`. Same for `if( !(false).equals(params.fasta))` - this can just be `if(params.fasta)`. Don't need to test the absolute value of the boolean - it's `false` by default and so will be true if set to anything else on the command line.

Reply:
ps. Maybe @pditommaso has a better "best practice" syntax for this: Channels created from existing files if supplied or from a process if not supplied.